[TorchTidy] Adding support for accessing strides and scalars

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #83009
* __->__ #80072

Differential Revision: [D37571570](https://our.internmc.facebook.com/intern/diff/D37571570)

Gamrix committed:
[TorchTidy] Adding support for accessing strides and scalars

[ghstack-poisoned]

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"




[ghstack-poisoned]

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"




[ghstack-poisoned]

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"




[ghstack-poisoned]

robieta commented on torch/csrc/profiler/collection.cpp:
```diff
@@ -21,6 +21,8 @@ void InputOutputEncoder::push(c10::ArrayRef<const c10::IValue> values) {
     } else if (value.isScalar()) {
       tags_.emplace_back(Tag::Scalar);
+      // Scalars are small enough to store as IValues
+      ivalues_.emplace_back(value);
```
How does the perf of holding an IValue compare to storing the Scalar directly? In theory the latter should be cheaper, but if its not meaningfully different obviously IValue has a bunch of advantages.

Gamrix replied:
For trivially copyable datatypes (which includes all scalars), the only difference in performance between an `IValue` and a `Scalar` would be an extra check on the tag to see if it is trivially copyable, which should not be meaningfully different. 

Both of them are tagged unions.
https://github.com/pytorch/pytorch/blob/fe6aa0844466e5dd2669092eac5edde153108b28/aten/src/ATen/core/ivalue.h#L1124

robieta reviewed (commented):
I have a couple nits, but overall looks good. Can you run the overhead benchmark to A/B test and see what the additional info costs?
```
buck build @mode/opt-clang //caffe2/caffe2/fb/high_perf_models/pytorch/benchmark_framework_overheads:cpp_benchmark
./buck-out/gen/caffe2/caffe2/fb/high_perf_models/pytorch/benchmark_framework_overheads/cpp_benchmark --useOutOfPlace --kinetoInputShapes
```
(Which, by the way, we should totally open source.)

Gamrix commented on test/test_profiler.py:
```diff
@@ -1175,8 +1175,25 @@ def test_tensor_sizes(self):
+        # The second argument to the add gets promotoed to a zerodim Tensor
+        self.assertEqual(node.extra_fields.inputs.dtypes, ['float', 'double', 'Scalar'])
+        self.assertEqual(node.extra_fields.inputs.shapes, [[5, 5], [], []])
+        self.assertEqual(node.extra_fields.inputs.ivalues, [None, None, alpha])
```
I will pull the layout stuff into a separate diff, I need to expose the `TensorMetadata` struct to Python.

Gamrix commented on torch/csrc/profiler/collection.h:
```diff
@@ -274,6 +277,8 @@ class InputOutputEncoder final {
   AppendOnlyList<TensorMetadata, IO_ENCODER_DEFAULT_BLOCK_SIZE>
       tensor_metadata_;
   AppendOnlyList<int64_t, IO_ENCODER_DEFAULT_BLOCK_SIZE> tensor_sizes_;
+  AppendOnlyList<int64_t, IO_ENCODER_DEFAULT_BLOCK_SIZE> tensor_strides_;
```
I don't see any reason not to, but I will likely refactor into using SizesAndStrides as you suggested above.

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"




[ghstack-poisoned]

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"




[ghstack-poisoned]

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"




[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37571570).

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"


Differential Revision: [D37571570](https://our.internmc.facebook.com/intern/diff/D37571570)

[ghstack-poisoned]

Gamrix commented:
This is rewritten into a new set of commits that does this.

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"


Differential Revision: [D37571570](https://our.internmc.facebook.com/intern/diff/D37571570)

[ghstack-poisoned]

Gamrix commented:
Reopening this commit, as in #81824 we concluded we would not do SizesAndStrides.

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"


Differential Revision: [D37571570](https://our.internmc.facebook.com/intern/diff/D37571570)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37571570).

Gamrix commented:
We can see that the changes slightly speed up the code. This is likely due to the bulk copy being more efficient than what we had before.

Base: 
https://our.internmc.facebook.com/intern/aibench/details/1688022957
https://our.internmc.facebook.com/intern/aibench/details/1778300628
https://our.internmc.facebook.com/intern/aibench/details/3685288181
https://our.internmc.facebook.com/intern/aibench/details/2328188198
https://our.internmc.facebook.com/intern/aibench/details/3554131505 

1.2508 1.2545 1.2627
1.2528 1.2554 1.2605
1.2637 1.2696 1.2789


D37571570
https://our.internmc.facebook.com/intern/aibench/details/1216719728 
https://our.internmc.facebook.com/intern/aibench/details/2852921225 
https://our.internmc.facebook.com/intern/aibench/details/308408276 
https://our.internmc.facebook.com/intern/aibench/details/3785164820 
https://our.internmc.facebook.com/intern/aibench/details/1536294584

1.2249 1.2278 1.2366
1.2244 1.2297 1.2459
 1.2216 1.2269 1.2405

robieta reviewed (approved):
LGTM.

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"


Differential Revision: [D37571570](https://our.internmc.facebook.com/intern/diff/D37571570)

[ghstack-poisoned]

Gamrix committed:
Update on "[TorchTidy] Adding support for accessing strides and scalars"


Differential Revision: [D37571570](https://our.internmc.facebook.com/intern/diff/D37571570)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37571570).
